cr-examples/onnx/src/main/java/oracle/code/onnx/OnnxRuntime.java

```diff
@@ -0,0 +1,535 @@
+            setInterOpNumThreads(1);
+        }
+
+        public void setInterOpNumThreads(int numThreads) {
+            try {
+                checkStatus(setInterOpNumThreads.invokeExact(sessionOptionsAddress, numThreads));
+            } catch (Throwable t) {
+                throw wrap(t);
+            }
+        }
+    }
+
+    public sealed interface OrtTypeInfo {
+    }
+
+    public final class OrtTensorTypeAndShapeInfo implements OrtTypeInfo {
+
+        private final MemorySegment infoAddress;
+
+        private OrtTensorTypeAndShapeInfo(MemorySegment infoAddress) {
+            this.infoAddress = infoAddress;
+        }
+
+        public long getDimensionsCount() {
+            try {
+                return retLong(getDimensionsCount.invokeExact(infoAddress, ret));
+            } catch (Throwable t) {
+                throw wrap(t);
+            }
+        }
+
+        public Tensor.ElementType getTensorElementType() {
+            try {
+                return Tensor.ElementType.fromOnnxId(retInt(getTensorElementType.invokeExact(infoAddress, ret)));
+            } catch (Throwable t) {
+                throw wrap(t);
+            }
+        }
+
+        public long getTensorShapeElementCount() {
+            try {
+                return retLong(getTensorShapeElementCount.invokeExact(infoAddress, ret));
+            } catch (Throwable t) {
+                throw wrap(t);
+            }
+        }
+
+        public TensorShape getShape() {
+            long dims = getDimensionsCount();
+            var shape = arena.allocate(JAVA_LONG, dims);
+            try {
+                checkStatus(getDimensions.invokeExact(infoAddress, shape, dims));
+            } catch (Throwable t) {
+                throw wrap(t);
+            }
+            return new TensorShape(shape);
+        }
+    }
+
+    public final class SequenceTypeInfo implements OrtTypeInfo {
+
+        private final MemorySegment infoAddress;
+
+        private SequenceTypeInfo(MemorySegment infoAddress) {
+            this.infoAddress = infoAddress;
+        }
+    }
+
+    public final class MapTypeInfo implements OrtTypeInfo {
+
+        private final MemorySegment infoAddress;
+
+        private MapTypeInfo(MemorySegment infoAddress) {
+            this.infoAddress = infoAddress;
+        }
+    }
+
+    private MemorySegment retAddr(Object res) {
+        checkStatus(res);
+        return ret.get(ADDR_WITH_ADDR, 0);
+    }
+
+    private int retInt(Object res) {
+        checkStatus(res);
+        return ret.get(JAVA_INT, 0);
+    }
+
+    private long retLong(Object res) {
+        checkStatus(res);
+        return ret.get(JAVA_LONG, 0);
+    }
+
+    private String retString(Object res) {
+        checkStatus(res);
+        return ret.get(ADDR_WITH_STRING, 0).getString(0);
+    }
+
+    private void checkStatus(Object res) {
+        if (!res.equals(MemorySegment.NULL) && res instanceof MemorySegment status) {
```

Comment:
This is another example of "hardwired logic". There's a struct definition for this:

```
struct OrtStatus {
  OrtErrorCode code;
  char msg[1];  // a null-terminated string
};
```

And the code is optimistically assuming that this layout will hold (across different versions of the API and across different platforms). Ok for a prototype of course, but that's another thing jextract might help you with.

Reply:
I agree the code is hard-wired to the ONNX library, however any significant changes in the API (more than reordering of the functions in the struct) will break the compatibility anyway. 
Jextracted code is also hard-wired to the specific library version.